cleverhans/picklable_model.py

```diff
@@ -354,3 +354,76 @@ def fprop(self, x):
 
     def get_params(self):
         return []
+
+
+class Add(Layer):
+    """
+    A Layer that adds a function to its input.
+    The function to add is specified in terms of multiple layers, just like
+    in the MLP class.
+    The Add layer is useful for implementing residual networks.
+    """
+
+    def __hash__(self):
+        return hash(id(self))
+
+    def set_input_shape(self, shape):
+        self.input_shape = shape
+        shapes = {"input": shape}
+        for layer in self.layers:
+            layer.set_input_shape(shapes[layer.parent])
+            shapes[layer.name] = layer.get_output_shape()
+        self.output_shape = shapes[self.layers[-1].name]
+
+    def __init__(self, layers):
+        super(Add, self).__init__()
+
+        self.layer_names = []
+        self.layers = layers
+
+        for i, layer in enumerate(self.layers):
+            if layer.parent is None:
+                if i == 0:
+                    layer.parent = "input"
+                else:
+                    layer.parent = layers[i - 1].name
+            if hasattr(layer, 'name'):
+                name = layer.name
+            else:
+                name = layer.__class__.__name__ + str(i)
+                layer.name = name
+            self.layer_names.append(name)
+
+    def get_params(self):
+        out = []
+        for layer in self.layers:
+            out = ordered_union(out, layer.get_params())
+        return out
+
+    def fprop(self, x, set_ref=False):
```

Comment:
Could you document was `set_ref` is for? I missed that in the `MLP` class as well. I assumed it is used to save a reference to the Python object for serialization w/ Pickle but correct me if I'm wrong.

Reply:
This was actually for reference batch norm / virtual batch norm, which we don't really use anymore. I just took it out.